core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java

```diff
@@ -16,276 +16,58 @@
+            aggregationUpdater = AtomicReferenceFieldUpdater.newUpdater(
+            AbstractStreamMessage.class, CompletableFuture.class, "aggregation");
 
-    @Override
-    public final void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
-                                SubscriptionOption... options) {
-        requireNonNull(subscriber, "subscriber");
-        requireNonNull(executor, "executor");
-        requireNonNull(options, "options");
-
-        final SubscriptionImpl subscription = new SubscriptionImpl(this, subscriber, executor, options);
-        final SubscriptionImpl actualSubscription = subscribe(subscription);
-        if (actualSubscription != subscription) {
-            // Failed to subscribe.
-            failLateSubscriber(actualSubscription, subscription);
-        }
-    }
+    @Nullable
+    private volatile CompletableFuture<?> aggregation;
 
     /**
-     * Sets the specified {@code subscription} for the current stream.
-     *
-     * @return the {@link SubscriptionImpl} which is used in actual subscription. If it's not the specified
-     *         {@code subscription}, it means the current stream is subscribed by other {@link Subscriber}
-     *         or aborted.
+     * Creates a new instance.
      */
-    abstract SubscriptionImpl subscribe(SubscriptionImpl subscription);
+    protected AbstractStreamMessage() {}
 
     @Override
-    public final CompletableFuture<Void> whenComplete() {
-        return completionFuture;
-    }
-
-    /**
-     * Callback invoked by {@link Subscription#request(long)} to add {@code n} to demand.
-     */
-    abstract void request(long n);
-
-    /**
-     * Callback invoked by {@link Subscription#cancel()} to cancel the stream.
-     */
-    abstract void cancel();
-
-    /**
-     * Invoked after an element is removed from the {@link StreamMessage} and before
-     * {@link Subscriber#onNext(Object)} is invoked.
-     *
-     * @param obj the removed element
-     */
-    protected void onRemoval(T obj) {}
-
-    static void failLateSubscriber(SubscriptionImpl actualSubscription, SubscriptionImpl lateSubscription) {
-        final Subscriber<?> actualSubscriber = actualSubscription.subscriber();
-        final Subscriber<?> lateSubscriber = lateSubscription.subscriber();
-        final Throwable cause = abortedOrLate(actualSubscriber);
-
-        if (lateSubscription.needsDirectInvocation()) {
-            handleLateSubscriber(lateSubscriber, cause);
-        } else {
-            lateSubscription.executor().execute(() -> {
-                handleLateSubscriber(lateSubscriber, cause);
-            });
+    public <U> CompletableFuture<U> aggregate(AggregationOptions<T, U> options) {
+        final ByteBufAllocator alloc = options.alloc();
+        final SubscriptionOption[] subscriptionOptions = alloc != null ? POOLED_OBJECTS : EMPTY_OPTIONS;
+        if (!options.cacheResult()) {
```

Comment:
Question) Is it possible that a user `aggregates` with different `AggregationOptions`?
e.g.
```
streamMessage.aggregate(options1); // with cache
streamMessage.aggregate(options2); // without cache
```

In this case, should we explicitly throw an exception?

Reply:
That was the most considering point.
Because of that, I redesigned the API multiple times while implementing it.
Originally, I tries to compare two options because it contains a function.
Identity comparison makes nonsense because if an aggregation happens inside of Armeria, users may not get the same instance of the option.

- For `StreamMessage`. It is an opt-in feature. The cache option is not exposed to users. They needs to enable the option using `AggregationOptionsBuilder()` on purpose.
- For `HttpRequest` and `HttpResponse`. The cache option is automatically enabled when `aggregate()` method is called. So most users don't need to see the option.

> In this case, should we explicitly throw an exception?

The aggregation with `option2` will eventually see `IllregalStateException` as the stream was subscribed already.